File: src/main/java/net/projecteuler/java/Problem124.java
package net.projecteuler.java;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import net.projecteuler.java.util.Number;

/**
 * Determining the k<sup>th</sup> element of the sorted radical function.
 *
 * @author vuser
 *
 */
public class Problem124 {

	public static void main(String ... args) {
		assert calculate() == 21417 : "failed";
	}

	private static int calculate() {
		List<Map<Long, Integer>> list = new ArrayList<Map<Long,Integer>>();

		// Insert one
		Map<Long, Integer> one = new HashMap<Long, Integer>();
		one.put(1L, 1);
		list.add(one);

		// Prime factorizations
		for (long i = 2; i <= 100000; i++) {
			Map<Long, Integer> primeExponent = Number.primeExponent(i);
			list.add(primeExponent);
		}

		// Comparator used by sort
		Comparator<Map<Long, Integer>> comparator = new Comparator<Map<Long,Integer>>() {

			/**
			 * Calculates the radicals of the two numbers based on the primes exponents.
			 * If the two radicals are equal the two numbers are compared.
			 */
			@Override
			public int compare(Map<Long, Integer> pe1, Map<Long, Integer> pe2) {

				int n1Rad = 1;
				int n2Rad = 1;

				int n1 = 1;
				int n2 = 1;

				for (Map.Entry<Long, Integer> e : pe1.entrySet()) {
					n1Rad *= e.getKey();
					n1 *= Math.pow(e.getKey(), e.getValue());
				}
				for (Map.Entry<Long, Integer> e : pe2.entrySet()) {
					n2Rad *= e.getKey();
					n2 *= Math.pow(e.getKey(), e.getValue());
				}
				if (n1Rad == n2Rad) {
					return n1 - n2;
				}

				return n1Rad - n2Rad;
			}
		};

		Collections.sort(list, comparator);

		int n = 1;
		for (Map.Entry<Long, Integer> e : list.get(9999).entrySet()) {
			n *= Math.pow(e.getKey(), e.getValue());
		}

		return n;
	}
}

File: src/main/scala/net/projecteuler/scala/Problem124.scala
package net.projecteuler.scala

import scala.collection.JavaConversions._
import scala.collection.mutable._
import java.lang.{ Long => JLong }
import java.lang.{ Integer => JInt }

import net.projecteuler.java.util.Number

/**
 * Determining the k<sup>th</sup> element of the sorted radical function.
 */
object Problem124 {

    def main(args: Array[String]): Unit = {

        var list = new ListBuffer[Map[JLong, JInt]]
        // Insert one
        list += Map(JLong.valueOf(1) -> 1);

        // Do the prime factorization
        for (i <- 2 to 100000) {
            list += mapAsScalaMap(Number.primeExponent(i))
        }

        // Sorts the list using the custom comparator
        var n = 1D
        for ((key, value) <- list.sortWith(compare)(9999)) {
            n *= Math.pow(key.toDouble, value.toDouble)
        }
        assert(n == 21417)

    }

    /**
     * Custom sorting function
     */
    def compare(pe1: Map[JLong, JInt], pe2: Map[JLong, JInt]) = {
        var n1Rad = 1L;
        var n2Rad = 1L;

        var n1 = 1D;
        var n2 = 1D;

        for ((key, value) <- pe1) {
            n1Rad *= key
            n1 *= Math.pow(key.toDouble, value.toDouble)
        }

        for ((key, value) <- pe2) {
            n2Rad *= key
            n2 *= Math.pow(key.toDouble, value.toDouble)
        }

        if (n1Rad == n2Rad)
            n1 < n2
        else
            n1Rad < n2Rad
    }
}